Hs 3015 as

aslobodeniuk opened the pull request.

ylatuya committed:
audiotracksink: create a new audio sink using the AudioTrack API

Fixes #HS-3016

ylatuya committed:
amcvideodec: add support for tunneled video playback

Fixes #HS-3017

ylatuya committed:
android: move the MediaFormat bindings to the androidjni library

ylatuya committed:
android: add a new API to query for tunneled playback

Fixes: #HS-3018

ylatuya committed:
audiotracksink: make the AudioTrack sink secondary

ylatuya committed:
amc: fix listing of tunneled and secure codecs

ylatuya committed:
amc: improve logging for audio session id

ylatuya committed:
amc: fix the property name used for tunneled playback

palegre-fluendo committed:
fix identation

ylatuya committed:
Add more logging

ylatuya committed:
audiotracksink: implement unlock

aslobodeniuk committed:
amc: push empty buffers in tunneled mode

ylatuya committed:
amc: push a dummy buffer going to PAUSED for preroll

aslobodeniuk committed:
amcvideo: fix deadlock on stopping

jmcatena commented:
As agreed during the daily, I am creating a branch HS-3027 based in this one to incorporate necessary fixes. Please do not merge this to sdk-10.0.23-hls before HS-3027 is merged into this.
Please let me know any concern.

jmcatena committed:
Fix adaptive mode

Adaptive playback means that the decoder can change the frame size on the fly,
so that no destruction is necessary to reconfigure.

Now if adaptive is enabled, we restart the amc codec only if format changed
more than the frame size.

The maximum frame size was being set with the current frame size. It needs to
be the maximum frame size to be decoded, but in dash we cannot know beforehand.
Now it is left unset, and verified to work with hikeen board. Documentation is
not clear if this is allowed, although leads to think so by tagging this as
optional and a hint for the codec to optimize resource allocation. In case
we find issues with other codec implementations, we'd need to set it from the
maximum size supported by the codec, at least when the maximum size cannot
be known.

This fixed the black gaps when the stream format changed in adaptive mode. It
still can happen if adaptive is not supported (unlikely in our targets), or when
the encoding changes. To fix those cases we cannot invalidate pushed buffers as
we are doing now. But that would be a more complex change that does not need to
be addressed urgently because this one solved the issue reported.

jmcatena committed:
Set the max frame size for adaptive with the maximum supported by amc codec

jmcatena committed:
Merge pull request #73 from fluendo/HS-3027-3

HS-3027-3

palegre-fluendo committed:
Get amc decoders list with a given feature

We can obtain a list of decoders name with one of these features:
adaptive-playback
tunneled-playback
secure-playback

We also convert the name of the decoder to gst plugin name.
OMX.qcom.video.decoder.avc.secure -> amcviddec-omxqcomvideodecoderavcsecure

jmcatena committed:
Improve important traces and readability

Tunneling and adaptive are new features and we need these traces to help
diagnosing potential defects.

jmcatena committed:
Merge pull request #75 from fluendo/HS-3028

Improve important traces and readability

aslobodeniuk commented:
just a reminder, that this branch has build problems, when we're building for the first time. Something related to accessibility of gstamcmacro.h . Maybe it's not a problem actually.

aslobodeniuk commented on sys/androidmedia/gstamcvideodec.c:
```diff
@@ -490,6 +451,20 @@ create_src_caps (const GstAmcCodecInfo * codec_info, gboolean direct_rendering)
+gst_amc_video_dec_push_dummy (GstAmcVideoDec * self)
+{
+  GstBuffer *buf = gst_buffer_new ();
+  GstCaps *caps = gst_caps_new_simple ("video/x-amc", NULL);
```
This is a small, but still a performance fall: we're creating empty "video/x-amc" caps each time (it adds several malloc/free), where we could just create it once.

aslobodeniuk commented on sys/androidmedia/gstamcvideodec.c:
```diff
@@ -490,6 +451,20 @@ create_src_caps (const GstAmcCodecInfo * codec_info, gboolean direct_rendering)
+  GstBuffer *buf = gst_buffer_new ();
+  GstCaps *caps = gst_caps_new_simple ("video/x-amc", NULL);
+
+  gst_pad_set_caps (GST_VIDEO_DECODER (self)->srcpad, caps);
```
This line is unacceptable: we are triggering new "set caps" proccess with each new buffer.

aslobodeniuk replied:
yes, this line has to be resolved, it's just caused by copying code in a hurry. I did it actually..

palegre-fluendo committed:
Merge pull request #74 from fluendo/HS-3039-2

Hs 3039 2

aslobodeniuk commented on gst-libs/gst/androidjni/gstjniamcutils.c:
```diff
@@ -0,0 +1,294 @@
+#include "gstjniamcutils.h"
+
+const gchar *
+gst_jni_amc_video_caps_to_mime (GstCaps * caps)
```
"const GstCaps * caps" would be nice

aslobodeniuk commented on gst-libs/gst/androidjni/gstjniamcutils.c:
```diff
@@ -0,0 +1,294 @@
+}
+
+gchar *
+gst_jni_amc_get_tunneled_playback_decoder_name (GstCaps * caps, gint width,
```
"const GstCaps * caps" would be nice

aslobodeniuk commented on gst-libs/gst/androidjni/gstjniamcutils.c:
```diff
@@ -0,0 +1,294 @@
+      "(Ljava/lang/String;)Z");
+
+  GST_ERROR
+      ("methods: get_capabilities_for_type_id %d\tget_name_id %d\tis_encoder_id %d\tis_feature_supported %d",
```
do we really need this trace? this seems to be a "development one"

palegre-fluendo replied:
Of course it is not critical, we can remove it or set DEBUG level.

aslobodeniuk committed:
amc: don't set caps to the pad with each buffer

fixes tunneling mode

jmcatena commented on gst-libs/gst/androidjni/gstjniamcutils.c:
```diff
@@ -0,0 +1,104 @@
+      return "video/mp4v-es";
+    else if (mpegversion == 1 || mpegversion == 2)
+      return "video/mpeg2";
+  } else if (strcmp (name, "video/x-h263") == 0) {
```
style: mixed braces and no braces with single statement blocks.

jmcatena commented on gst-libs/gst/androidjni/gstjnimediacodeclist.c:
```diff
@@ -0,0 +1,136 @@
+
+  env = gst_jni_get_env ();
+
+  jobject tmp = (*env)->FindClass (env, "android/media/MediaCodecList");
```
Nit: we have already a function to avoid repeating this block everywhere: j_find_class
Also, gstamc keeps a cache of used classes and methods so that they do not have to be obtained everytime. That probably should belong to a lower level common jni support library, but anyway it's there to be used.

aslobodeniuk replied:
we actually also have gst_jni_get_class, which is the same as j_find_class

jmcatena commented on gst-libs/gst/androidjni/gstjniutils.h:
```diff
@@ -79,4 +79,93 @@ _type gst_jni_get_##_name##_field (JNIEnv *env, jobject obj, jfieldID fieldID);
+   J_CALL_<TYPE> (), J_CALL_STATIC_<TYPE> () - first parameter is a variable
+   to write to, if it's not J_CALL_VOID () or J_CALL_STATIC_VOID ().
+   If exception occured, it jumps to "error" label, and the variable
+   is kept untouched.
```
NIT: Jumping to labels from macros makes usage of this macros very tricky. We should consider another way to communicate the success/failure, like a bool variable as first parameter.

aslobodeniuk replied:
I have the way-around opinion: this pattern of having an "error" label in each function should be used in all the code we write, because it simplifies the code, and at the same way forces the code to check and log every unexpected case.

jmcatena commented on sys/androidmedia/gstamc.c:
```diff
@@ -60,16 +54,26 @@ static gboolean ignore_unknown_color_formats = FALSE;
-gst_amc_format_get_jstring (GstAmcFormat * format, const gchar * key,
-    jstring * value);
 
 /* Global cached references */
```
NIT: We may want to have the cached jni classes and methods in a more general library, maybe gstgniutils.
Instead of definining each class in a root structure, we could beter use nested structures resembling the java namespaces.

jmcatena commented on sys/androidmedia/gstamc.c:
```diff
@@ -1499,6 +1344,19 @@ get_java_classes (void)
+      "intValue", "()I");
+
+  android_range.klass = j_find_class (env, "android/util/Range");
+  GST_ERROR ("range.klass=%p", android_range.klass);
```
Delete this line, I left it from testing and is not needed.

jmcatena commented on sys/androidmedia/gstamc.c:
```diff
@@ -2135,7 +1928,7 @@ scan_codecs (GstPlugin * plugin)
 
     codec_info_class = (*env)->GetObjectClass (env, codec_info);
-    if (!codec_list_class) {
+    if (!codec_list) {
```
I think the intention was to check codec_info_class, not codec_list.

jmcatena commented on sys/androidmedia/gstamc.c:
```diff
@@ -3257,6 +3048,19 @@ register_codecs (GstPlugin * plugin)
       else
         rank = GST_RANK_PRIMARY;
 
+      /* FIXME: this should be done looking at the codec feature and it
```
Important: In sdk we are raising or lowering codecs supporting tunneled rank by 1 depending on if tunneled is desired or not.
Lowering the rank seems unnecessary and sdk may need to alter rank by larger amount to compensate this.

ylatuya replied:
This is required because otherwise you would have 3 decoders with same rank and by default we want the regular decoder to be used

jmcatena commented on sys/androidmedia/gstamcvideodec.h:
```diff
@@ -58,6 +61,8 @@ struct _GstAmcVideoDec
   GstVideoCodecState *input_state;
   gboolean input_state_changed;
 
+  gboolean tunneled_playback_enabled;
```
Unused, can be safely removed, the one in _GstAmcCodec is used instead.

jmcatena commented:
I finished reviewing and I have an important concern:
Currently we are activating adaptive if the decoder supports it, regardless of tunneled or not.
But adaptive requires it to be configured to output to a surface, it does not work when output is byte buffers.
If we are using byte output in any use case, there are many more fixes to do, including those that cause black gaps and issues when changing format.
I suppose that it is not using byte buffer output even in non tunneled mode because adaptive works fine and it should not otherwise, but reviewing the code it's difficult to be sure, so maybe someone that know better can clarify and save more study of this already huge PR.
We cannot keep including big fixes in this PR, we need to merge what's done, but we need to understand what to expect from this state and what's pending if any.

jmcatena committed:
Remove unused field.

jmcatena reviewed (approved):
I have concerns expressed before, but these may be adressed if necessary in further PRs and are not a reason to block this one.

aslobodeniuk committed:
Merge pull request #76 from fluendo/HS-3015-as-fixes

amc: don't set caps to the pad with each buffer

turran committed:
Always include local paths before remote ones

aslobodeniuk committed:
amc: apply "const" to some function definitions

aslobodeniuk committed:
jni/amc: resolve existence of 2 same functions

aslobodeniuk committed:
jni/amc: indent to same codebase used

aslobodeniuk committed:
jni/amc: fix log output, add FIXME comment

aslobodeniuk committed:
amcvideodec: cache caps for dummy buffer

to avoid extra malloc/free calls
